### ISSUE TITLE
Improve or document testability by users

### ISSUE THREAD

**Issue by DannyBen** [piotrmurach/tty-prompt#139]
One of the problems I am facing with TTY::Prompt is the lack of formal way of testing interaction with the menu (at least to my knowledge).

Up until now, I was hacking it by capturing STDOUT and STDIN, and injecting keystrokes into it, for example, like can be seen [here](https://github.com/DannyBen/menu_commander/blob/master/spec/spec_mixin.rb)

This proved to be fragile, as evident in https://github.com/piotrmurach/tty-screen/issues/11.

I would love to have a formal way to test it, which would allow me to "inject" output to the prompt - regardless of what type it is.

Pseudo-code wise, I would love to `include TTY::SpecHelpers` in my rspecs, and then write something like:

```ruby
tty_send_input :down, :down, :enter do
  subject.execute ...
end
```

Or, by stubbing a single, known input handler:

```ruby
expect(TTY::Prompt.input_handler)
  .to receive(:get_input)
  .and_return [:down, :down, "hello", :enter]
```

**Comment by DannyBen** [comment]
I see, I see... Thank you. Looking forward to it.

**Comment by DanielVartanov** [comment]
Hey @DannyBen, I feel your pain, I'd been in your exact situation just recently and stumbled upon the same things. My case was also worsened by the fact that I had a multi-stage CLI app with multiple consequtive questions/choices. 

Basically we need a system where we can _feed_ stdin of an app with any content and expect it to just work.

As Piotr mentioned above I ended up with a way of achieving that without too much black magic. I am now in the process of convincing Piotr that this way of testing is worth being included into ttytoolkit family, please wish me luck :-)

**Comment by DannyBen** [comment]
HI @DanielVartanov , happy to hear something is in the works.

I consider myself a heavy TTY::* user, and have quite a few tools using it, and many that use TTY::Prompt in particular.

In my projects, I also pay attention to have efficient tests, so I also usually tend to avoid copy-pasting repetitive code. For this purpose, I developed [rspec_approvals](https://github.com/dannyben/rspec_approvals), which is an interactive approval testing - meaning, it will show you the output, you can approve/reject it, and it will ask you again if the output changes.

Why am I telling you this? In some of my specs, I simply create a YAML file with a) command to run and b) keyboard input to feed, and let the specs execute them. I then approve the output interactively (using TTY::Prompt of course), and the specs will "let me know" if any output changes from the approved content in the future.

- [Here is an example of such YAML file](https://github.com/DannyBen/menu_commander/blob/master/spec/menu_commander/examples.yml)
- [Here is how it is being fed to rspec](https://github.com/DannyBen/menu_commander/blob/master/spec/menu_commander/examples_spec.rb#L30-L32)
 - [Here is the spec_helper that provides the interactive "feeding" of stdin](https://github.com/DannyBen/menu_commander/blob/master/spec/spec_mixin.rb)

This approach works, and as you can see (by the spec helper code), it requires only a handful lines of code. Zero black magic of any kind. The problem with it, is since it is not "officially" supported, any change (such as [a recent one in TTY::Screen](https://github.com/piotrmurach/tty-screen/issues/11)) could potentially break it.

Long story short, if there is any repo or gist where I can take a look and possibly contribute to this effort, let me know.

**Comment by piotrmurach** [comment]
@DanielVartanov I'm 100% convinced. Apologies if I made you think that I don't approve of this idea or don't want to make it part of TTY toolkit. I thought I was voicing encouragement and supporting your experiments. I really do like the look of where the experiments are going! I juggle many things so for me it's more matter of finding time than anything.

So to bring everyone on the same page, here are my current thoughts:

* TTY toolkit needs a dedicated testing solution, therefore I'm keen to pull together any helpers that may be reusable into a single gem.
* I'd love this to power the unit and integration testing for the `tty` gem itself. For some time, I wanted to improve and provide out-of-the-box experience to make testing much simpler.
* There are many testing frameworks but I don't believe we need to provide any specific assertions. The implementation could be exposed as a module that can be included in any test class.
* No hard dependency on tty gems to keep the interface as flexible as possible. The core principle of `tty` gems is that they are easy to mix & match with other libraries. However, I envisage a specific module that provides some niceties for testing `tty-prompt` like sending keystroke names. In the spirit of what Danny has suggested.

I will setup repository soon so we can collaborate.

**Comment by DanielVartanov** [comment]
@piotrmurach that's awesome to hear! And please don't apologize, it's job of an offerer to convince the maintainer, as there's enough people out there who just bully maintainers to do what they want.
Anyway, I too can devote time only once per few days, but I am using soon-to-be tty-testing in a real project so I constantly get experience of what works and what only sounds good. So, please take your time, no need to rush.

@DannyBen awesome lib! Love to community developing artful console tools and apps. Let me share a couple of examples of what I am trying to do here: https://gist.github.com/DanielVartanov/89e08ca436eb73cbaeaae1912f3133cd

Perhaps in the particular case you shared this will make little difference for you, but next time when you do a multi-question CLI app with subprocess (called via tty-command) which must comply with what you feed to stdin too, this approach will help you a lot.

**Comment by piotrmurach** [comment]
@DannyBen I refactored tests and made `TTY::Prompt::Test` public to help test prompt interactions. This should help you make your tests more robust and immune against changes. The `tty-test` gem will make testing much nicer and allow for more complex interactions but I see this as a reasonable intermediate step given that I use it internally to test `tty-prompt`.

**Comment by DannyBen** [comment]
Looks cool, thanks! Will give it a try.

**Comment by DanielVartanov** [comment]
Today I'm ready to present a first proper draft of the testing tool for tty toolkit
https://github.com/DanielVartanov/tty-testing

@DannyBen please check whether it fits your purposes

> The tty-test gem will make testing much nicer

Right, I see it should've been named `tty-test`, let's rename it after I transfer the ownership to you.

**Comment by DannyBen** [comment]
Wow, so many TTY gifts in one weekend!
I will take a look soon, thanks for the notification.

**Comment by DannyBen** [comment]
Ok. My initial observations for the `TTY::Prompt::Test` approach.

1. There seem to be a small mistake in the README. We need to require 'tty/prompt/test' and not `tty-prompt/test`.

2. The example [provided in the README](https://github.com/piotrmurach/tty-prompt#4-testing) is helpful, but might confuse novice users. In reality, the `prompt` instance is not the subject under test . The `prompt` instance will normally be tucked away inside some subject, assuming the developer considered dependency injection.

3. I have created a [sample gist](https://gist.github.com/DannyBen/6ad4ade97961856dc63efe3138046fcc) showing how real life specs might look like ("real life" for anyone NOT testing the prompt itself of course).

4. Even with this minimal example, it took me a while to get it working due to the fact that providing invalid input (like missing an Enter), causes it to just get stuck waiting for input behind the scenes. I suspect that testing complex interactions with this might not be easy. This problem exists in any other testing method that exists until now, so its not worse, just also not better.

**Comment by DannyBen** [comment]
In regards to the new `TTY::Testing` gem - I am sorry, I really don't get it. Things there seem to be way too complicated, and confusing to me.

The [examples](https://github.com/DanielVartanov/tty-testing#1-basic-usage) are creating an app wrapper, and a prompt. In reality, we already have a CLI app, and the prompt is called from within it. We do not want to stub or mock the entire app. Just its input and output streams.

I am not sure if its just that the examples are confusing, or the entire philosophy of the testing gem is confusing, so forgive me if I missed the point.

**Comment by DanielVartanov** [comment]
@DannyBen humm, I didn't think it might look confusing, thanks for a valuable feedback, I might need to recraft the examples.

So, basically your app is what to be wrapped by the wrapper, your app should accept `stdin` and `stdout` as parameters (it is a proper practice anyway) and then it look like this:

```ruby
let(:app) do
  TTY::Testing.app_wrapper do |stdin, stdout|
     MyCLIApp.new(stdin: stdin, stdout: stdout).run!
  end
end

it 'asks questions' do
  expect(app.output).to eq "What is your name?"
  app.input.puts "DannyBen"
  expect(app.output).to end_with "Hello, DannyBen!"
end
```

while the prompt is indeed created inside your `MyCLIAapp` class

**Comment by DannyBen** [comment]
Thanks. The above example is clearer, but...

> your app should accept stdin and stdout

This requires changing of the application, just for the sake of testing. Not ideal.
Also, I never saw this kind of implementation in any other rspec plugin / helper.

Usually, plugins (such as this one, which is a plugin for TTY testing) does one of these:

1. Adds matchers - like `redirect_to` in
```ruby
expect(subject).to redirect_to(path)
```
2. Adds convenience methods - like `get` in
```ruby
expect { get '/path' }.to be_successful
```
3. Adds stubbing conventions, which is what I would expect more or less from a TTY testing gem.
```ruby
expect(subject).to receive(prompt).and_return(mock_prompt)
# or 
allow(TTY::Testing.input).to receive(:input).with("Hello", :down, :enter)
```
4. Adds "test drivers" - like Rails is doing for things like mail deliveries, or browsing the app - this is also a good option for TTY testing in my view - think setting the TTY Prompt "driver" to "test mode" before running tests.

In my view, there is either one or two things that just need stubbing. Everything else is noise. I want to:

1. Either stub `MyApp#prompt` and have it return a test prompt, on which I can inject stdin and examine stdout/stderr (which is what the `tty/prompt/test` approach provides in a sense).
2. Or, stub the input and output of TTY itself, or in other words, stub the "driver" - as is common in other testing related test plugins (in Rails). Example:
```ruby
# in spec_helper.rb
TTY::Prompt.driver = TTY::Prompt::Test

# in any test
TTY::Prompt.input = ["Hello", :down, :enter]
MySubject.run_however_i_see_fit
expect(TTY::Prompt.output).to eq "Hello Jax"
```

For reference, this is similar to `expect(ActionMailer::Base.deliveries).not_to be_empty` in Rails, when testing mailers.

I guess the three of us here have three different use cases / needs / philosophies... :)

**Comment by DanielVartanov** [comment]
Hi Danny,

Before I reply, let me show something I did in a fork to your gist. No offense or patronising, just a way to demonstrate things:

At first, the app is interacting with the user with input and output, but for tests at jus returns plain value as a string in the method `#run`, that's a bit odd, console apps are meant to be tested via user input and output. I tool a liberty to update the code so that it gets tested via its inputs and ouputs, please take a look when you have time:

https://gist.github.com/DanielVartanov/19d5cfe8f4cb9215af40c14229ed48ae

I also added an example of how it would be tested via `tty-test(ing)`.

**Comment by DanielVartanov** [comment]
Now, to your points. A huge, fundamental and a principal difference between `TTY` gems and `Rails` is that the former is a tool and the latter is a _framework_. With frameworks you write code where you were told to  -- normally in special methods of descendants of special classes (controllers, models, mailers etc).

> This requires changing of the application, just for the sake of testing. Not ideal.

The concept is called [Dependency Injection](https://en.wikipedia.org/wiki/Dependency_injection) and under most circumstances it is the best  way to organise any app for the sake of testability, maintainability, loose coupling etc. In short, it is a very good and common practice of organising things.

Not surprisingly `Rails` does _exactly that_ when you want it to `expect(subject).to redirect_to(path)`, in tests it feeds controllers with mock HTTP requests while in development/production environment it feeds it with real HTTP requests received from the network socket. Exactly the same principle is applied here in `TTY` except for it is visible to the user since, again, `TTY` gems are _tools_, not _frameworks_.

**Comment by DannyBen** [comment]
> The concept is called Dependency Injection

I know what dependency injection is, thank you.

In rspec, it is done by stubbing methods to return a different object, or by mocking - not by requiring the app under test to have something (stdout/stdin) that it doesnt normally need.

I am sorry, but if this is the direction this gem takes, I feel like testing in any other method - like hijacking stdin/stdout or like using the newly available `TTY::Prompt::Test` might be better.

I will try to create some better rspec example - and maybe even a "helper" - to demonstrate my point.

**Comment by DanielVartanov** [comment]
Hey Danny,

Sorry, not tried to be patronising or something, I genuinely thought you just didn't know about DI. 

> In rspec, it is done by stubbing methods to return a different object, or by mocking

RSpec does not really define that, a necessity to expose every substituted entity as a public method and and prohibiting the tested class to use the entity in any other way by via that public method is not what RSpec does. It is definitely a way to test with RSpec, but just one of the many.

> I will try to create some better rspec example - and maybe even a "helper" - to demonstrate my point.

Great! :+1: That would be most helpful, thanks in advance.

**Comment by DannyBen** [comment]
... and I really meant "thank you for sharing knowledge" when I said thank you, but now when reading it I can see why it sounded like I'm annoyed... :) - I do appreciate knowledge sharing.

In any case - since both the testing gem and the exposure of `TTY::Prompt::Test` are not yet providing me with the "zen" of testing I expect (https://github.com/piotrmurach/tty-prompt/issues/148), I believe I have something that will probably work well for many TTY users testing in rspec, and it is **much** simpler than anything we all seen so far.  Hint: It will bypass input streams.

I figured this is analogous to the difference between being clever and being smart - instead of being clever and trying to fix the problem, I want to try and be smart and avoid it altogether.

I will post a proof of concept after I try it in (my-) real world scenarios.

**Comment by piotrmurach** [comment]
Thank you both for this discussion, it is interesting to see how you wish to approach testing CLIs. This is definitely needed if we are to have an approach that works and is intuitive to use. Minimal and intuitive APIs are something I really care about. 

@DannyBen Thank you for giving `TTY::Prompt::Test` a try and providing feedback. I totally agree with your points and in particular the weak points of this approach. I wanted to take a 'first' step and provide an 'intermediate way' that will provide some form of direction for testing prompt interactions. Possibly a help and improvement over global mocking of standard inputs. 

It may be preferable and more flexible to replace the `TTY::Prompt::Test` with `TTY::Prompt` and demonstrate how to use `:input` and `:output` parameters. At least then nothing will seem like magic and the user will have full knowledge and full control of what is going on. I also want to keep this testing framework agnostic, so I didn't want to use `subject` and `before` helpers. We could possibly have `rspec` and `minitest` examples.

As for my 'philosophy' of testing. Since Rails has been used as comparison, I will try to draw some parallels. Please bear in mind that my perspective is skewed towards the `tty` and the `teletype` executable. 

Having said that, for me these are related:

```
request <=> stdin
response <=> stdout
```

In `tty` and `teletype` the equivalent of `controller` is a concept of `command` that when executed takes in some standard input and produces 'potentially' some standard output. There may be some side effects like creation of files etc. But to all intents and purposes, I want to provide a way and a guide for unit testing commands. These commands may use few `tty-*` components which as Daniel mentioned will be 'wired' using standard in and out. However, at this stage, we don't care whether `tty-prompt` is used or some other components from other gems. 

In my experience, providing mocked input and output tends to provide a robust solution for doing uni testing of CLIs. I often use IO-like object as I'm not interested in interactions but the content that the user inputs and observers afterwards. Stubbing interactions using doubles would focus on implementation details and not the business logic, which to me is the input and output cycle. More concretely:

```ruby
# => initialize all dependencies including `tty-*` components
cmd = SomeCommand.new
# a more functional approach to mocking interactions similar to the Phoenix framework
# where each action receives all the connection information
cmd.execute(stdin: ..., stdout: ...) 
```

We could also approach it from the perspective of stubbing all the interactions within the command. So, if a command uses, let's says 5 `tty` components, we would stub all of them and provide canned responses. However, I worry this may lead to brittle tests knowing too much about the command's internals and as such may 'miss' interactions between components instead of 'truly' testing what happens. So in this respect, I'm more 'classicist' rather than 'mockist' because I don't like tying tests to the implementation and prefer to check actual state changes. In my view, the way to achieve this is by injecting standard input and output to the test subject whatever it may be. I'm not dead set on this but this has been the approach that served me well in all `tty` components and the open-source CLIs, [coinpare](https://github.com/piotrmurach/coinpare) being one of them. Please see [coins_spec](https://github.com/piotrmurach/coinpare/blob/master/spec/unit/coins_spec.rb). It may well be that we will have some 'hybrid' approach that uses stubs and mocked input and output. I don't know really, but I want to explore the best approach.

I'd like to have more discussion on this as we don't have a solution that matches expectations. Hence I must admit I always 'put off' thinking and creating a library to handle testing tty components. My current goal is to finish updating all the `tty` components(nearly there) so that I can remove the bundler `< 2.0` requirement in the `tty` gem itself. This is to say that we don't need to settle on any approach yet and I'm keen on exploring how `tty-test` may/should work so I could use it and suggest for testing `teletype` generated CLIs.

**Comment by DannyBen** [comment]
Thanks for the thoughtful reply @piotrmurach - one comment that I have, based on everything I know and have seen so far in the "tty testing" discussion: I am starting to get the feeling that there is **no need** for a specific testing gem. I feel that the TTY components that are responsible for providing input and output, should be slightly altered - similar to how you made the Test prompt public - in order to allow for better stubbing and tesatbility.

In my mind, it is similar to how each of us improve our own code when we face hard-to-test areas.

I believe that there could be still a place for a `TTY::Test` gem, or a `require 'tty/rspec` helper - but that the core solution lies within the TTY gem suite itself, and these additional gems would just be some syntactic sugar.

**Comment by DannyBen** [comment]
@piotrmurach @DanielVartanov - as promised, I am "opening the case wide open" with another approach.

I have created [this minimal sample repo](https://github.com/DannyBen/tty-prompt-spec-example) - with a readme and everything.

In a nutshell: We never reach a point where we need input/output streams. 

### Teaser:

```ruby
it "works" do
  ask "Matz"    # expect(subject.prompt).to receive(:ask).and_return "Matz"
  select "Ruby" # expect(subject.prompt).to receive(:select).and_return "Ruby"

  expect(subject.run).to eq "Matz likes Ruby"
end
```

**Comment by piotrmurach** [comment]
@DannyBen If I understand your idea well, your preference would be to add a set of custom assertions that would make it easier to provide 'canned responses'.  Curious to know how would your approach testing in more complex scenarios that Daniel is trying to address? In the latest 'unreleased' as of yet version, I added `:quiet` option that allows you to skip the standard output altogether. You can use this option to build a 'dynamic' menu with nested options (see [menu example](https://github.com/piotrmurach/tty-prompt/blob/master/examples/menu.rb)). How would you test-drive command that depends on the dynamic input which is not provided at the subject creation or any method call?

**Comment by DannyBen** [comment]
> If I understand your idea well, your preference would be to add a set of custom assertions

Well, not sure. The one thing I do know, is that I want to work *with* rspec, not against it.
So - if someone wants to mock the entire "prompt" object, let them. If they want to stub input and output, let them. 

However, if we can come up with some helper methods - analogous to `get '/path'` in rails, that will universally work for everybody, then why not.

> how would your approach testing in more complex scenarios

Personally, I never had to deal with CLI apps that have long and complex interactions. But, borrowing from other areas where I had to deal with complex code, and then test it - there is only one solution. Make it shorter and simpler by separating it to different methods.

Regardless of which will be the TTY testing tool - I think that it will not be a good idea to test long interactions as a whole. They must be separated, and then the tests will be easier.

Nevertheless, I understand that Daniel's experience and requirements may differ, and perhaps his use case requires a more elaborate testing framework. 

> How would you test-drive command that depends on the dynamic input

Not sure I am following. At the end of the day, any app-under-test ends up calling one of `TTY::Prompt`'s methods - so it can be stubbed just like anything.

Does the example menu you provided above demonstrate your dynamic menu? If so, I can take it and "spec it". If not - can you provide some example?

**Comment by DannyBen** [comment]
Well - I see the difficulty testing this dynamic example without actually having control over stdin - and BTW, I love this example and its `quiet` option!

Although I think it may still be possible to test such examples without `stdin` control, I am not sure it can be done intuitively and naturally enough.

So - perhaps go back one gear to the idea of having a single method somewhere in `TTY` that can be stubbed so that it will get an array representing the input.

Behind the scenes, I expect it to read like this:

```ruby
def tty_input(*keys)
  expect(TTY::InputHandler).to receive(:read).and_return(keys)
end
```

This should be able to handle special named keys - `:enter`, `:down` etc.

The test is expected to look something like this:

```ruby
describe CoolMenu do
  it "works" do
    # Increase coolness to 3 then exit
    tty_input 2, 1, 1, 1, 3
    subject.run
    expect(subject.cool).to eq 3
  end
end
```

This of course, will cause the menu to print to the screen, interleaving its output with the spec output - which is undesirable. This is why I really think that TTY needs to have a "test driver" which will swallow all the stdout that prompt itself causes (but NOT the other stdout prints).

```ruby
# spec_helper.rb
TTY::Prompt.driver = :test
```

For completeness, I converted the dynamic menu example to a more testable class, if anyone wants to try cracking this scenario.

<details>
  <summary>Show the CoolMenu class</summary>

```ruby
require "tty-prompt"

class CoolMenu
  attr_writer :cool

  def prompt
    @prompt ||= TTY::Prompt.new(quiet: true)
  end

  def cool
    @cool ||= 0
  end

  def run
    main_menu
  end

protected

  def main_menu(from = nil)
    name = prompt.select("Cool Menu") do |menu|
      menu.enum "."

      menu.choice "Coolness Status", :status_menu
      menu.choice "Manage Coolness", :manage_menu
      menu.choice "Exit", "exit"
    end

    next_menu(name, :main_menu)
  end

  def next_menu(name, from)
    if name.is_a?(Symbol)
      send(name, from)
    else
      eval(name)
    end
  end

  def status_menu(from)
    name = prompt.select("Coolness is at #{cool}" ) do |menu|
      menu.enum "."

      menu.choice "Back", from
      menu.choice "Exit", "exit"
    end

    next_menu(name, :status_menu)
  end

  def manage_menu(from)
    name = prompt.select("Coolness is at #{cool}.\nManage" ) do |menu|
      menu.enum "."

      menu.choice "Add coolness", :add_cool
      menu.choice "Back", :main_menu
      menu.choice "Exit", "exit"
    end

    next_menu(name, :manage_menu)
  end

  def add_cool(from)
    self.cool += 1
    next_menu(from, from)
  end
end

# Run with:
# CoolMenu.new.run

```
</details>

**Comment by DannyBen** [comment]
Alright - posting this as a new message, since the previous one was too long already.

I have updated the example repo to show more or less how stdin control might look like.

This spec: ([on github](https://github.com/DannyBen/tty-prompt-spec-example/blob/master/spec/app_spec.rb))

```ruby
require 'spec_helper'

describe App, :tty do
  context "without arguments" do
    it "asks for a name and favorite language" do
      tty_input "Matz", :enter, :down, :enter
      expect(subject.run).to eq "Matz likes Ruby"
    end
  end

  context "with --confirm" do
    it "also asks if I am sure" do
      tty_input "Matz", :enter, :down, :enter, 'y'
      expect(subject.run %w[--confirm]).to eq "Matz likes Ruby"
    end
  end

  context "with --print" do
    it "prints the output" do
      tty_input "Matz", :enter, :down, :enter
      expect { subject.run %w[--print] }
        .to output("==> Matz likes Ruby\n").to_stdout
    end
  end
end
```

Made possible by [this spec_helper](https://github.com/DannyBen/tty-prompt-spec-example/blob/master/spec/tty_rspec.rb)

**Comment by DanielVartanov** [comment]
A little side note: it's been a pleasure reading this thoughtful discussion.

As someone who's been deep in CLI testing just recently I've got only two things to amend:

1.

>  I am starting to get the feeling that there is no need for a specific testing gem

I surprisingly agree :-D 
In 90% of cases CLI apps are simple one-step feed-and-read experiences (`coinpare` mentioned above is a typical example of it), such apps definitely need no additional structures, libraries or tools. The whole thing about `tty-test(ing)` gem is to deal with multi-stage interactive apps with many many branches of possibilities, it's a heavy duty tool useful for 10% of apps at best.

2. I'd like to bring your guys attention to a problem of subprocesses which are often used in CLI apps, e.g. when one CLI app runs another CLI app (probably a system one) which's output is used in the main app output. Testing such things would require a proper file descriptor of an IO object passed to `TTY::Prompt`. Though, it is not an often case either.

**Comment by DannyBen** [comment]
I am glad we are all converging to a similar thought pattern.

> I'd like to bring your guys attention to a problem of subprocesses which are often used in CLI apps, e.g. when one CLI app runs another CLI app (probably a system one)

In this regard - when the CLI calls something else - I follow what I believe is considered a best practice in testing in general, not only CLI: 

There is no need to test the output of that subprocess, and there is not even a need to execute it. We are assuming it works, and that it has its own tests to prove it. All we need to do, is ensure that *our* CLI tool *calls it*. So: `expect(subject).to receive(:system).with(["--some", "args"])`

In other words - testing that subprocess is outside the scope of your own test suite, in my view.

**Comment by DannyBen** [comment]
## State of the (TTY Testing) Union

I feel we are all more or less close to reaching conclusions, so trying to summarize our situation here in order to materialize this discussion to its desired outcomes.

### Tester Types

There will probably be three testing scenarios, for different users and use cases (going from simple to complex):

1. Testers who design their tests and app to avoid the need to control STDIN.
2. Testers who want or need STDIN injection control.
3. Testers who need more than STDIN/STDOUT control.

### Implementation

#### Case 1: Stub TTY::Prompt methods

For case 1, we can easily stub `TTY::Prompt` methods, perhaps with the help of helpers that will be user-defined in the `spec_helper` (i.e. no need to add anything to TTY for this case).

Example: `expect(subject.prompt).to receive(:select).and_return("Pizza")`

#### Case 2: TTY::Prompt::Test

For case 2, we can stub STDIN. So far, the best approach for this was to have `TTY::Prompt::Test` available. If this solution can be polished (for example to avoid the `StringIO` hacks, as I believe was already done) - then this solution is also ready.

I am not mentioning STDOUT specifically, since a) it can be captured natively in rspec with `expect { subject.run }.to output("something").to_stdout` and b) the `TTY::Prompt::Test` already provides `.output.string`, so we can test the prompt output itself if desired.

#### Case 3: Complex Apps

For case 3, there is probably a need for the separate testing gem that allows for a more involved interaction with the app. Personally, I do not think this gem should be the main recommended testing approach for TTY users, since it will usually be an overkill.

### Next Steps

My proposal for next steps:

- [ ] Complete polishing of `TTY::Prompt::Test` or an alternative approach to allow stdin stubbing.
- [ ] Create a Markdown document describing and demonstrating the first two test approaches. This can either be put into the README or into a Wiki page, dealers choice. (I can help with this document).
- [ ] Complete the `TTY::Test(ing)` gem, and link to it from the above document.

**Comment by DanielVartanov** [comment]
I'd like to elaborate on subprocesses thingie at least to refer to this convo in future. Several points on the matter, just for the record  (i.e. not trying to argue further, everyone picks their own way of testing):

I.  Besides `Kernel#system` there are  many other ways to run a subprocess: `Process.spawn`, `fork/exec`, `Open3`, `tty-command` just to name a few. Tests shall test behaviour, not the implementation.

II. Implementation can switch between running a subprocess and doing the work itself, e.g. between running `curl` and using `libcurl/curb` (not the greatest example as curl's output isn't very much valuable, but I hope it illustrates the idea).

III. Stubbing and mocking are used to test a unit in isolation from other units, such tests are a great compliment to functional/acceptance tests, but they are never sufficient (they can pass green when the app does not work and fail when the app actually still works).

**Comment by DanielVartanov** [comment]
Hi Danny,

Great, I love your summary. Perhaps the very fact that we are having this conversation alone states that there are just different preferences in approaches to testing and tty-toolkit should not impose one upon the user.

I actually have an idea on polishing `TTY::Prompt::Test`, will try to write it up today in the evening, if I imagine everything correctly it will work exactly as you describe it in your examples.

**Comment by DanielVartanov** [comment]
After playing a bit with your ideas of `TTY::Prompt.driver = ...` and `prompt.input = %w[1 2 3]` I managed to pull it together, but since this approach is not my cup of tea I can't really assess whether it's any good

```ruby
TTY::Prompt.test_mode! # in spec_helper.rb

RSpec.describe "console app" do
  let(:prompt) { TTY::Prompt.new } # No TTY::Prompt::Test

  subject do
    name = prompt.ask("Name?")
    if prompt.yes?("Greetings?")
      prompt.say "Hello, #{name}!"
    end
  end

  specify do
    prompt.input = "Danny\nY\n"
    subject
    expect(prompt.output).to end_with("Hello, Danny!\n")
  end
end
```

Implementation and this ^^^ test being runnable as a single file is in [this gist](https://gist.github.com/DanielVartanov/8eac7a04fb77322f9422cdf1ed9d67f4)
(sorry implementation is quick-n-dirty, it's just a PoC that such DSL can work)

**Comment by DannyBen** [comment]
Thanks Daniel. You gotta love Ruby and the neat tricks it lets you do... :)

I added your example to the [demo repo](https://github.com/DannyBen/tty-prompt-spec-example) and invited you both.

I used the idea of a "driver" mostly for inspiration, since this is how Rails is doing it and in some cases it makes sense. I believe that in our case - now that `TTY::Prompt::Test` is shaping up to be (almost?) completely functional and available to end users - we can avoid that use of a driver.

The below code (also available as [test2 in the repo](https://github.com/DannyBen/tty-prompt-spec-example/blob/master/3-example-daniel/test2.rb)) can be achieved with this:

```ruby
require 'tty-prompt'
require 'tty/prompt/test'

class TTY::Prompt::Test
  def input=(keys)
    keys.each { |key| input << key }
    input.rewind
  end
end

RSpec.describe "console app" do
  let(:prompt) { TTY::Prompt::Test.new }

  subject do
    name = prompt.ask("Name?")
    if prompt.yes?("Greetings?")
      prompt.say "Hello, #{name}!"
    end
  end

  specify do
    prompt.input = ["Danny", "\n", "Y", "\n"]
    subject
    expect(prompt.output.string).to end_with("Hello, Danny!\n")
  end
end
```

Advantages:

1. We do not contaminate `TTY::Prompt` itself
2. We utilize an already existing `TTY::Prompt::Test` which is specifically designed for testing.
3. The line `  let(:prompt) { TTY::Prompt::Test.new }` expresses intent more clearly than `  let(:prompt) { TTY::Prompt.new }` - in fact, this is our "setting of a test driver" right here.

In my mind - if we can:

1. Add this `input=` method to `TTY::Prompt::Test`, and
2. Make it accept symbols for special keys - `:enter :down :pgup`, then this portion of our experiment can be over, and we can have a nice approach for testing (at least for the 90% apps we discussed).

**Comment by DannyBen** [comment]
And just to clarify some things about the "test driver" approach, so that we are all on the same page.

First, I am also not a big fan of this approach. I dislike it in Rails, for two reasons:

1. It is a little bit "magic" - in Rails you even not necessarily know that it is configured, since it is hidden inside some config/environment file that came like this as a default.
2. It causes your specs to look less like DSL, since it contains a lot of code calling namespaced modules (e.g. `expect(ActionMailer::Base.deliveries).not_to be_empty`). Since Rspec invested TONS of resources in making your specs read naturally, this kind of syntax breaks the natural reading pace.

That said, I believe if we have a proper `TTY::Prompt::Test` - it is a **better approach** than having a test driver, since it *is* the test driver in a way, and since `TTY::Prompt::Test` already exists and 90% operational.

For completeness, I want to describe - in my view - what is the core concept behind a test driver.

To me - when using a test driver - it means that you configure that "tool" to avoid performing its standard operation, and instead, write its artifacts in a single, known location. 

In `ActionMailer`, this location is `ActionMailer::Base.deliveries`:

```ruby
something_that_sends_mail.run
expect(ActionMailer::Base.deliveries).not_to be_empty
```

In `ActiveJob` the situation is both worse (since it is more complicated behind the scenes) and better (since it is more DSLish to the user) and rspec needs to provide a specific matcher for it:

```ruby
expect { SomeJob.perform_later }.to have_enqueued_job
```

Yuck for both.

So if we were to use a driver approach (not saying we should), it should have looked like this:

```ruby
# spec_helper

# Enable test mode in a way that can be disabled mid-flight if needed
TTY::Prompt.test_mode = true

# spec

# Inject input, to a predetermined static location
it "works" do
  TTY::Test.input = ["Piotr", :enter, "y", :enter, :down, :enter]
  subject.run
  expect(TTY::Test.output).to match "Hello Piotr"
end

# Alternatively - if we borrow from ActiveJob's test approach:
it "works" do
  TTY::Test.input = ["Piotr", :enter, "y", :enter, :down, :enter]
  expect { subject.run }.to have_tty_output("Hello Piotr")
end
```

**I know these are horrible** - but this is the "driver approach" in my mind, which we don't need.

**Comment by piotrmurach** [comment]
Thank you for all your comments! I read all them and hopefully haven't missed anything.

To me, the discussion touches on two main subjects, one of how to test `tty-prompt` interactions within some context and one a bit wider of testing CLIs(e.i. running external processes that may or may not deal with user input). This is great but as more immediate step I'd probably choose to focus on `tty-prompt` as suggested at the start of this thread and split the discussion and continue a wider approach to testing of CLIS and `tty` gems in the `tty` project itself. 

#### Case 1: Stub TTY::Prompt methods

This is a viable solution, though I'm not convinced about recommending this in docs as an approach for testing classes that use `TTY::Prompt` to perform their business logic. One of the 'weaknesses' that I see is the fact that the 'canned' responses are predetermined. The choices and options to any prompt methods maybe dynamic and a result of some other computation. Secondly, stubbing 'separates' implementation. It's not about testing `tty-prompt` itself but ensuring that, for example, input conversion happens and, for instance, returns a boolean answer. Lastly, stubbing adds a lot of 'noise' in a test especially when the user deals with few prompts in a given context of a class. I have therefore reservations about making this as a recommended approach.

#### Case 2: TTY::Prompt::Test

I use this approach in other `tty` gems like `tty-file`. The benefits of using `TTY::Prompt::Test` are many. You don't need to stub any prompt methods. If a subject under test calls many prompts, the test doesn't need to 'know' about them. This means that the developer is free to use whatever the prompt as long as the end result works. This separates the test from the implementation details and focuses it more on business logic. What binds the different prompts is the fact that standard input is mocked. You control test execution with different inputs. For example, using `select` prompt may actually create many execution paths through a codebase which depend on different inputs.

To be more precise, in `tty-file` is use `expand` prompt which allows a user to choose action when there is a conflict with a file. Each option in the menu triggers different logic and needs to be tested separately. For example, the test for file collision with overwriting logic looks like this:

```ruby
it "displays collision menu and overwrites" do
  test_prompt = TTY::Prompt::Test.new
  test_prompt.input << "\n"
  test_prompt.input.rewind
  allow(TTY::Prompt).to receive(:new).and_return(test_prompt)
```

After the setup that tells the prompt to overwrite file, I run the action and observe output:

```ruby
  file = path_factory.call("README.md")
  TTY::File.create_file(file, "# Title", verbose: false)

  expect {
    TTY::File.create_file(file, "# Header", verbose: true)
  }.to output(/\e\[31mcollision\e\[0m  #{file}/).to_stdout_from_any_process
  expect(test_prompt.output.string).to match(/Overwrite #{file}\?/)
```

And do some final sanity check:

```ruby
 expect(File.read(file)).to eq("# Header")
```

What I like about this test is that I can swap `expand` prompt for another one and the test wouldn't need to change at all provided that the input triggers the same behaviour.  There are some 'rough' edges in this approach. One, the `TTY::Prompt::Test` could expose much nicer methods. I'd like to suggest couple of methods:

```ruby
test_prompt.type_in "input" # an actual input
test_prompt.press :enter  # a generic way to named concepts like :enter, :space, :up
```

Then common scenarios may be replaced with:

```ruby
  test_prompt.input << "\n"
  test_prompt.input.rewind
  # instead do this
  test_prompt.press_enter
```

To me, this is where Daniel's approach using `Fibers` may actually yield(sorry;) great benefits. If anything, it could be used to power more 'natural' flow to inputs and reverse the logic of the previous test:

```ruby
  test_prompt = TTY::Prompt::Test.new
  allow(TTY::Prompt).to receive(:new).and_return(test_prompt)

  file = path_factory.call("README.md")
  TTY::File.create_file(file, "# Title", verbose: false)

  expect {
    TTY::File.create_file(file, "# Header", verbose: true)
   test_prompt.press_enter # much nicer!
 } ....
```

In general, this approach avoids any mysterious configuration options and permits granular usage. You can have a file with many tests and only few of them focus on the prompt interactions.

#### Case 3: Complex Apps

I think this is more about the level of abstraction and the type of tests you wish to write rather than being 'complex'. Whether you call them unit or integration tests, the aim is to get the most value out of a single test. To me, the more outside-in approach for testing CLIs is more preferable especially in the context of auto-generated tests. In other words, I'm talking about writing `capybara` like tests to exercise a given scenario. This can be a single command or a few commands working together. For example, in `tty` gem the new command has a lot of logic and single public `execute` method and I think it would be tedious and rather difficult to write unit tests that would stub and mock many calls inside. There are way too many inputs/flags and external commands to consider stubbing them all. These commands like Daniel have mentioned, need to be run to affect a real change in the system. In the case of the new command, create an entire CLI structure.

Example of how I think this could work:

```ruby
cmd = Command.new # subject under test with single `execute` method
cmd_session = TTY::Test::Session.new(cmd) # injects mocked inputs and output and hiding all the 'wiring' details

cmd_session.execute(...) # forwards all calls to the underlying cmd

cmd_session.type_in "input" # input sent to whatever listens on the mocked standard in, could be `gets` method
cmd_session.press_enter # send "\n" to whatever still listens

# check some intermediate results

cmd_session.press :up, :up, :enter

# check some intermediate results
```

I'd love to see some PRs for the `TTY::Prompt::Test` that tackle what we have discussed so far and make the API easier to use and more intuitive.

ps @DanielVartanov I released `tty-prompt` and I will soon release other gems.

**Comment by Papipo** [comment]
Taking a seat, since I need to test a Thor app that uses `TTY::Prompt` under the hood to display a menu and I don't know how to test it.
